### PR TITLE
feat: allow for overriding project asset handling

### PR DESCRIPTION
**Pull request by sergiusens**
- Move executable setting to one location
- Test for default behavior in relation to hooks
- Test with new asset handler in relation to hooks

- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `tox run -m lint`?
- [ ] Have you successfully run `tox run -e test-py310`? (supported versions: `py39`, `py310`, `py311`, `py312`)

-----

### CODE REVIEW

**Comment by sergiusens** [comment]
The Intention is to later do something like this in the package service:

```python
    @override
    def write_metadata(self, path: pathlib.Path) -> None:
        """Write the project metadata to metadata.yaml in the given directory.

        :param path: The path to the prime directory.
        """
        meta_dir = path / "meta"
        meta_dir.mkdir(parents=True, exist_ok=True)

        self.metadata.to_yaml_file(meta_dir / "snap.yaml")

        assets_dir = self._get_assets_dir()
        setup_assets(
            self._project,
            assets_dir=assets_dir,
            project_dir=self._services.lifecycle.project_info.project_dir,
            prime_dir=path,
            meta_directory_handler=meta_directory_handler,
        )

    @property
    def metadata(self) -> snap_yaml.SnapMetadata:
        """Get the metadata model for this project."""
        return snap_yaml.get_metadata_from_project(
            self._project, self._services.lifecycle.prime_dir, arch=self._build_for
        )


def meta_directory_handler(assets_dir: pathlib.Path, prime_dir: pathlib.Path):
    built_snap_hooks = prime_dir / "snap" / "hooks"
    project_dir_hooks = assets_dir / "hooks"

    hooks_meta_dir = prime_dir / "meta" / "hooks"
    hooks_meta_dir.mkdir(parents=True, exist_ok=True)

    if built_snap_hooks.is_dir():
        for hook in built_snap_hooks.iterdir():
            meta_dir_hook = hooks_meta_dir / hook.name
            # Remove to always refresh to the latest
            meta_dir_hook.unlink(missing_ok=True)
            meta_dir_hook.hardlink_to(hook)

    # Overwrite any built hooks with project level ones
    if project_dir_hooks.is_dir():
        for hook in project_dir_hooks.iterdir():
            meta_dir_hook = hooks_meta_dir / hook.name
            meta_dir_hook.unlink(missing_ok=True)

            try:
                meta_dir_hook.hardlink_to(hook)
            except OSError as os_error:
                # Cross device link
                if not os_error.errno == 18:
                    raise
                shutil.copy(hook, meta_dir_hook)
```

File: snapcraft/parts/setup_assets.py [hunk 1]
```diff
@@ -33,23 +33,35 @@
 
 
 def setup_assets(
-    project: models.Project, *, assets_dir: Path, project_dir: Path, prime_dir: Path
+    project: models.Project,
+    *,
+    assets_dir: Path,
+    project_dir: Path,
+    prime_dir: Path,
+    meta_directory_handler: Optional[Callable[[Path, Path], None]] = None,
```
Comment (lengau) [inline snapcraft/parts/setup_assets.py:41]: I appreciate the solid typing here and the docstring explaining how the handler is called.

**Review by lengau — APPROVED** [review]
Clear and concise. Thanks!

File: snapcraft/parts/setup_assets.py [hunk 2]
```diff
@@ -33,23 +33,35 @@
...
 ) -> None:
     """Copy assets to the appropriate locations in the snap filesystem.
 
     :param project: The snap project file.
     :param assets_dir: The directory containing snap project assets.
     :param project_dir: The project root directory.
     :param prime_dir: The directory containing the content to be snapped.
+    :param meta_directory_handler: handler for hooks built through the lifecycle,
```
Comment (mr-cal) [inline snapcraft/parts/setup_assets.py:49]: Pedantic: This is not intuitive (I cannot figure out what this function does based on the signature and docstring).

I think this should be more like `:param hook_handler: callback that overrides default behavior for handling hooks`

**Review by mr-cal — APPROVED** [review]
Minor comment on the parameter name and docstring but the design looks good

File: snapcraft/parts/setup_assets.py (hunk 2 above)
Comment (sergiusens) [reply snapcraft/parts/setup_assets.py:49]: it also handles gui, I see my comment is stale, will fix!

**Comment by sergiusens** [comment]
the new ros failures might be due to #4619

**Pull request merged**